renormalizer/mps/gs.py

```diff
@@ -134,365 +96,436 @@ def optimize_mps(mps: Mps, mpo: Mpo, omega :float =None) -> Tuple[List, Mps]:
-                    precond = lambda x, e, *args: x / (hdiag - e + 1e-4)
-                    
-                    e, c = davidson(
-                        hop, cguess, precond, max_cycle=100,
-                        nroots=nroots, max_memory=64000
-                    )
-                    # if one root, davidson return e as np.float
-
-                #elif algo == "arpack":
-                #    # scipy arpack solver : much slower than pyscf/davidson
-                #    A = scipy.sparse.linalg.LinearOperator((nonzeros,nonzeros), matvec=hop)
-                #    e, c = scipy.sparse.linalg.eigsh(A, k=nroots, which="SA", v0=cguess)
-                #    # scipy return numpy.array
-                #    if nroots == 1:
-                #        e = e[0]
-                #elif algo == "lobpcg":
-                #    precond = lambda x: scipy.sparse.diags(1/(hdiag+1e-4)) @ x
-                #    A = scipy.sparse.linalg.LinearOperator((nonzeros,nonzeros),
-                #            matvec=hop, matmat=hop)
-                #    M = scipy.sparse.linalg.LinearOperator((nonzeros,nonzeros),
-                #            matvec=precond, matmat=hop)
-                #    e, c = scipy.sparse.linalg.lobpcg(A, np.array(cguess).T,
-                #            M=M, largest=False)
-                elif algo == "primme":
-                    precond = lambda x: scipy.sparse.diags(1/(hdiag+1e-4)) @ x
-                    A = scipy.sparse.linalg.LinearOperator((nonzeros,nonzeros),
-                            matvec=hop, matmat=hop)
-                    M = scipy.sparse.linalg.LinearOperator((nonzeros,nonzeros),
-                            matvec=precond, matmat=hop)
-                    e, c = primme.eigsh(A, k=min(nroots,nonzeros), which="SA", 
-                            v0=np.array(cguess).T, OPinv=M,
-                            method="PRIMME_DYNAMIC", 
-                            tol=1e-6)
-                else:
-                    assert False
-                logger.debug(f"use {algo}, HC hops: {count}")
+                            raw_cguess = tensordot(mps[cidx[0]], ms, axes=1)
+                    cguess.append(asnumpy(raw_cguess)[qnmat == mps.qntot])
+
+            guess_dim = np.sum(qnmat == mps.qntot)
+            cguess.extend(
+                [np.random.rand(guess_dim) - 0.5 for i in range(len(cguess), nroots)]
+            )
+            e, c = eigh_iterative(mps, qnmat, ltensor, rtensor, cmo, omega, cguess)
+
+        # if multi roots, both davidson and primme return np.ndarray
+        if nroots > 1:
+            e = e.tolist()
+        logger.debug(f"energy: {e}")
+        micro_iteration_result.append(e)
+
+        cstruct = cvec2cmat(cshape, c, qnmat, mps.qntot, nroots=nroots)
+
+        # store the "optimal" mps (usually in the middle of each sweep)
+        should_store_res_mps: bool = False
+        if last_opt_e_idx is not None:
+            if method == "1site":
+                res_mps_idx = last_opt_e_idx
```

Comment:
I think the `res_mps_idx` is still not correct even for 1-site algorithm. For example: 4 sites,0,1,2,3 sweeping from right to left. The micro_iteration_result is [e3,e2,e1,e0]. if the minimum energy is e1, the `res_mps_idx=2`, however the correct res_mps_idx for the next left-to-right sweep is `imps=1`.

Reply:
Maybe replace the current type of `micro_iteration_result` with a dictionary `{(active_site):e,...}` will make the bug easy to solve.

Reply:
I see your point here. The culprit is `micro_iteration_result.append(e)` even though the iteration is not from left to right.

Reply:
Yes, you are right.